feat(chore): add lint using CI

mvorisek opened the pull request:
## Description

Unified CS improves readability, prevents hidden mistakes and enforce unified contribution standards. And in a long term, it also simplifies merge conflicts.

The CS rules are based on industry standard `eslint-config-airbnb-base`, but all rules might be too hard to fix at once. This PR focuses on whitespace CS which can be autofixed /wo having to review each line manually.

related https://github.com/fomantic/Fomantic-UI/issues/29

nicely integrated with Github Actions like:

![image](https://user-images.githubusercontent.com/2228672/195988285-21a68de4-a29b-4604-8418-e9088be5a105.png)

prudho commented:
Yes, code linting is definitely something we need to implement. I started using myself eslint at work and this is something i don't regret at all. But... 😅 

We cannot implement it, and already use it to enforce merge requests. I think that first we need to define which rules we want to use and the level of error it'll raise. Then we can do some merge requests to fix the code. But we can't fix all the code in one PR, maybe a PR per component.

Once this is done, then we can enforce lint check trough the CI pipeline.

mvorisek commented:
I have updated the PR, instead of using `airbnb-base` as a base eslint config (with possible side effects) I added the most important whitespaces rules only to match general coding standards. With whitespace changes the code behaviour is not changes and this PR is safe to merge.

In another PR, more eslint rules can be added.

Also don't worry about merge conflicts (from changes based on code before this PR), they can be easily resolved, just cherry pick the eslint rules and run `npm run lint-fix`, you can then integrate the changes as usual.

lubber-de reviewed (commented):
As the PR is basically ok now, i just comment instead of requesting changes as i would like to hear opinion of other maintainers. See my comments below

mvorisek commented on examples/assets/library/jquery.min.js:
```diff
@@ -1,4 +1,4 @@
-!function(a,b){"object"==typeof module&&"object"==typeof module.exports?module.exports=a.document?b(a,!0):function(a){if(!a.document)throw new Error("jQuery requires a window with a document");return b(a)}:b(a)}("undefined"!=typeof window?window:this,function(a,b){var c=[],d=c.slice,e=c.concat,f=c.push,g=c.indexOf,h={},i=h.toString,j=h.hasOwnProperty,k={},l=a.document,m="2.1.4",n=function(a,b){re ...
-return M.access(a,b,c)},removeData:function(a,b){M.remove(a,b)},_data:function(a,b,c){return L.access(a,b,c)},_removeData:function(a,b){L.remove(a,b)}}),n.fn.extend({data:function(a,b){var c,d,e,f=this[0],g=f&&f.attributes;if(void 0===a){if(this.length&&(e=M.get(f),1===f.nodeType&&!L.get(f,"hasDataAttrs"))){c=g.length;while(c--)g[c]&&(d=g[c].name,0===d.indexOf("data-")&&(d=n.camelCase(d.slice(5)) ...
-void 0===c?d&&"get"in d&&null!==(e=d.get(a,b))?e:(e=n.find.attr(a,b),null==e?void 0:e):null!==c?d&&"set"in d&&void 0!==(e=d.set(a,c,b))?e:(a.setAttribute(b,c+""),c):void n.removeAttr(a,b))},removeAttr:function(a,b){var c,d,e=0,f=b&&b.match(E);if(f&&1===a.nodeType)while(c=f[e++])d=n.propFix[c]||c,n.expr.match.bool.test(c)&&(a[d]=!1),a.removeAttribute(c)},attrHooks:{type:{set:function(a,b){if(!k.ra ...
+!function(a,b){'object'==typeof module&&'object'==typeof module.exports?module.exports=a.document?b(a,!0):function(a){if(!a.document){throw new Error('jQuery requires a window with a document');}return b(a)}:b(a)}('undefined'!=typeof window?window:this,function(a,b){var c=[],d=c.slice,e=c.concat,f=c.push,g=c.indexOf,h={},i=h.toString,j=h.hasOwnProperty,k={},l=a.document,m='2.1.4',n=function(a,b){ ...
```
fixed - `/examples/assets/library` added to eslint excluded files, no foreign library should be transformed

mvorisek commented on .github/workflows/ci.yml:
```diff
@@ -1,9 +1,19 @@
 name: CI
-on: [pull_request]
+on: [push, pull_request]
```
Having here `push` implicitly is done in a lof of project, any push will run CI.

This makes experiments easier and also runs CI for development/master and other release branches.

If ok with you, I prefer to keep `push`.

mvorisek commented on .eslintrc.js:
```diff
@@ -0,0 +1,30 @@
+        'test/coverage',
+    ],
+    rules: {
+        indent: ['error', 4, { SwitchCase: 1 }],
```
I personally highly prefer 4 spaces as 2 spaces requires more eye stress to track the current indentation level and also forces you visually to not write deeply nested code /wo using methods.

So when normalizing the CS now, I propose 4 spaces. But if most other maintainers will require 2 spaces, I would not/cannot object.

lubber-de commented:
I took a look at the complete files to get myself convinced about the 4 indent 😉  and  i recognized the linter does sometimes not implement the 4 indent correctly, which
needs (manual?) adjustments, otherwise the new indent is making it worse than before at those codelines, because it gets confusing

For example dimmer.js, look at the semicolon, which terminates the var statement
https://github.com/fomantic/Fomantic-UI/blob/9074205244996ae83a06273e5364ca2be190fca4/src/definitions/modules/dimmer.js#L38

It is now looking like that
![image](https://user-images.githubusercontent.com/18379884/205279833-784a14c9-5225-484c-9cbc-1b933aa117a5.png)

while it was and is supposed to look like that (same indent as the var statement)
![image](https://user-images.githubusercontent.com/18379884/205279987-24600811-3921-4448-a780-8a3f760c4372.png)

mvorisek commented:
Maybe we need another eslint rule for semicolon, I will have a look on a solution for it.

mvorisek committed:
Unify CS and add CI linter

mvorisek committed:
add semi rule

mvorisek committed:
fix semi indent

mvorisek commented:
The solution is to run eslint `semi: ['error', 'never']` once and then again. There is no explicit config of indentation for semicolon in eslint.

lubber-de commented:
Well, when adding the semicolon to the end of the last line, why the indention at all then (in such cases)?

![image](https://user-images.githubusercontent.com/18379884/205399964-f577ab7f-3f07-46c3-b10a-01ad32509ac2.png)


Also, you added the rule to add a trailing comma in object literals. The linter idea behind that is to easily exchange/adding/removing single lines without the need to also add another comma to the previous attribute notation.
This is the same for such use cases where the semicolon has its own line.

So: Adding the trailing comma but also adding a semicolon to the last entry is not consistent anymore (IMHO even worse)
I would really prefer to have the semicolon being indented as it was before. This clearly shows where the indentation block ends as it is now.

Now we have the discussion i basically want to avoid ;) And now imagine new contributors trying to help and we are going to argue about "correct" whitespacing :) 

Don't get me wrong, I am not against the linting. But indentin the semicolons would be the cleaner approach. Yes, this possibly needs a (onetime) manual adjustment if there is no 100% automatic solution.

mvorisek commented:
> Well, when adding the semicolon to the end of the last line, why the indention at all then (in such cases)?

chained method call indentation is defacto standard (at least has always been in all codebase I have worked with) and it makes sense, as

```
$module
    .callOne()
    .callTwo();
$module.callOne();
```

is IMHO visually more readable and more compact than

```
$module
.callOne()
.callTwo()
;
$module.callOne();
```

> Don't get me wrong, I am not against the linting. But indentin the semicolons would be the cleaner approach. Yes, this possibly needs a (onetime) manual adjustment if there is no 100% automatic solution.

You are right here, I have reviewed the changes and there are at least two usecases with semicolon on a NL:

a) ternary:

![image](https://user-images.githubusercontent.com/2228672/205403485-5d8b1777-6de9-4940-9f6b-af16ffe971d9.png)

I have never seen semicolon on NL for them, also no reason for easier diffing, as ternary operator is always binary.

b) var declaration:

![image](https://user-images.githubusercontent.com/2228672/205403696-3e7bf5af-99f5-4a94-9b40-dcad7d8ea498.png)

I will fix these manually 👍

exoego commented:
For the record, I am positive on introducing linter.
I believe it helps both contributors and maintainers.
I defer to lubber on styling preference since he is the most active maintainer.

lubber-de commented:
> ```
> $module
>     .callOne()
>     .callTwo();
> $module.callOne();
> ```
> 
> is IMHO visually more readable and more compact than
> 
> ```
> $module
> .callOne()
> .callTwo()
> ;
> $module.callOne();
> ```
> 

My point basically was about comparison to the trailing comma rule.
If you want to add a .callThree() you now have to remove the semicolon from .callTwo() as well

> I will fix these manually 👍

Thx 😊

mvorisek committed:
fix semi with (var|const|let) regex

((?:var|const|let)\n[^;]+^( +)    [^;\n]+);\n +

mvorisek committed:
add no-trailing-spaces rule

regex: ;\n +\n + eslint fix /w review

mvorisek committed:
manual review/fixes of regex replace

mvorisek committed:
add brace-style rule

mvorisek committed:
manual fixes

mvorisek committed:
add keyword-spacing rule

mvorisek committed:
add space-in-parens rule

mvorisek committed:
add padded-blocks rule

mvorisek committed:
add comma-spacing rule

mvorisek committed:
add space-before-function-paren rule

mvorisek committed:
add no-multiple-empty-lines rule

mvorisek committed:
add space-infix-ops rule

mvorisek committed:
add no-multi-spaces rule

mvorisek committed:
add padding-line-between-statements rule

mvorisek committed:
add spaced-comment rule /w manual changes

mvorisek committed:
review all comments indent

^( *)(//.*|/\*(?:(?!\*/)[\S\s])+\*/)\n\1

mvorisek commented:
This PR is complete.

I have updated the PR title/description to stress it focuses solely on whitespaces. There is no single functional change.

I have reviewed all manual changes by generating unminified `semantic.js` before and after this PR and then applied eslint with the following <details>
  <summary>config</summary>
```
module.exports = {
    env: {
        browser: true,
        es6: true,
    },
    extends: [
    ],
    parserOptions: {
        ecmaVersion: '2020',
        sourceType: 'module',
    },
    ignorePatterns: [
        '/dist',
        '/examples/assets/library',
        '/test/coverage',
        '/test/helpers',
    ],
    rules: {
        indent: ['error', 4, { SwitchCase: 1 }],
        'keyword-spacing': ['error'],
        'space-in-parens': ['error'],
        'comma-spacing': ['error'],
        'space-before-function-paren': ['error', {
            'named': 'never',
            'anonymous': 'always',
            'asyncArrow': 'always',
        }],
        'space-infix-ops': ['error'],
        'padding-line-between-statements': ['error', {
            blankLine: 'always',
            prev: '*',
            next: ['continue', 'break', 'export', 'return', 'throw'],
        }],
        'spaced-comment': ['error', 'always', {
            line: {
                markers: ['/'],
                exceptions: ['-', '+'],
            },
            block: {
                markers: ['!'],
                exceptions: ['*'],
                balanced: true,
            },
        }],
        'padded-blocks': ['error', 'never'],
        'no-multiple-empty-lines': ['error', {
            max: 0,
            maxBOF: 0,
            maxEOF: 0,
        }],
        'no-multi-spaces': ['error'],
        'key-spacing': ['error'],
        'brace-style': ['error'],
        curly: ['error'],
        quotes: ['error', 'single', { avoidEscape: true }],
        'comma-dangle': ['error', {
            arrays: 'always-multiline',
            objects: 'always-multiline',
            functions: 'never',
            imports: 'always-multiline',
            exports: 'always-multiline',
        }],
        semi: ['error', 'never'],
        'no-extra-semi': ['error'],
        'no-trailing-spaces': ['error'],
    },
    reportUnusedDisableDirectives: true,
};
```
</details>

to remove all autofixed whitespace differencies as much as possible. The changes made look correct.

lubber-de commented:
Thanks for doing this.
I am sorry to say, but i still don't agree to the inconsisty of having trailing commas in object literals, but also putting the semicolon to the end of  chained methods. What one solves is re-invented by the other

The idea for the trailing comma rule is
```javascript

// bad code as per lint rule
// When i want to add another object property, i have to change 2 lines (add the comma to c: 2 and my new line)
var a = {
    b: 1,
    c: 2
}

// 2 lines changed. Bad practice as per eslint rule
var a = {
    b: 1,
    c: 2,
    d: 3
}

//  but lint rule wants to only change one line which is considered goot practice
var a = {
    b: 1,
    c: 2,
    d: 3,
}
```

I am basically totally fine with the above rule :)

But now:
You changed from 
```javascript

$module
    .on('bla',function(){})
    .on('bla',function(){})
    .on('bla',function(){})
;
```

to

```javascript

$module
    .on('bla1',function(){})
    .on('bla2',function(){})
    .on('bla3',function(){});
```

Regardless of the opinion based readability, we now have invented the same bad practice problem as the trailing comma force was solving by the previous rule before.
If i want to add another chain function to the end of the chain i have to remove the semicolon in the previous line.

```javascript
// 2 lines need  to be changed
$module
    .on('bla1',function(){})
    .on('bla2',function(){})
    .on('bla3',function(){})
    .on('bla4',function(){});


// while the current code would allow for only adding the additional chain line
// 2 lines need  to be changed
$module
    .on('bla1',function(){})
    .on('bla2',function(){})
    .on('bla3',function(){})
    .on('bla4',function(){})
;

```

That's what i meant by inconsisty.

lubber-de commented on src/definitions/behaviors/form.js:
```diff
@@ -8,2070 +8,2029 @@
+
+                            return (
+                                $field.filter('#' + identifier).length > 0 ||
+              $field.filter('[name="' + identifier + '"]').length > 0 ||
+              $field.filter('[data-' + metadata.validate + '="' + identifier + '"]').length > 0
```
indent is completely missing

mvorisek replied:
fixed - eslint does not indent lines starting/ending with an operator

mvorisek committed:
semi on a NL for chained & multiline calls

(^( *)    \.\w.*\(.*);

mvorisek committed:
semi on a NL for chained & multiline calls II.

(^( *)    \.\w.*\(.*(\n +(?! )(?!;|\.))[\s\S]+?^\2    \}\));

mvorisek committed:
-1 ident for "$allModules.each"

located by previous regex

mvorisek committed:
semi on a NL for chained & multiline calls III.

using previous regex replace run repeatable until zero matches /w manual review

mvorisek committed:
fix multiline operator not indented

 (-|\+|%|/|\*|\|\||&&)(?<!( |\n) \*)$

mvorisek commented:
I was able to craft regexes to fix semicolon on a NL for chained calls. Is there anything left?

mvorisek committed:
add quote-props rule

mvorisek committed:
add object-curly-spacing rule

mvorisek committed:
add space-before-blocks rule

mvorisek committed:
add arrow-parens rule

mvorisek committed:
add object-shorthand rule

mvorisek committed:
add array-bracket-spacing rule

mvorisek committed:
add dot-notation rule

mvorisek committed:
fix tool import/order and import/no-useless-path-segments rules

mvorisek committed:
add one-var-declaration-per-line rule

mvorisek committed:
add function-paren-newline rule

mvorisek committed:
fix tasks manually

mvorisek committed:
add computed-property-spacing rule

mvorisek committed:
add object-curly-newline rule

mvorisek committed:
add operator-linebreak rule

mvorisek committed:
add comma-style rule

mvorisek commented:
PR is done, I added more whitespaces rules to match AirBnB CS (basically the rules are copied from their ruleset). The hash of minified `semantic.min.js` file before and after this PR is unchanged, ie. there is no functional change.

Later, I plan to create another PR which will fix `var` -> `let` etc., which can have side effects.

lubber-de reviewed (approved):
LGTM! Thanks for doing this and your patience 😉
